[SYCL][Doc] math functions added to bfloat16 ext

JackAKirk opened the pull request:
This SYCL 2020 extension proposal proposes adding bfloat16 support to the `fma`, `fmin`, `fmax` and `fabs` SYCL floating point math functions.

cc @dkhaldi @AlexeySotkin 

Would this extension proposal be of interest to you?

Blocked by https://github.com/intel/llvm/pull/5393

gmlueck commented:
Rather than creating a separate extension, would it make sense to combine this with SYCL_INTEL_bf16_conversion, and create a new experimental extension named "sycl_ext_intel_bfloat16" (or "sycl_ext_oneapi_bfloat16")?  It seems like it would be easier for applications if all this bfloat16 functionality was tied to a single feature-test macro.

We also need to decide if these are "intel" or "oneapi" extensions.  It makes no sense to have a "oneapi" extension layered on top of an "intel" extension.

JackAKirk commented:
> Rather than creating a separate extension, would it make sense to combine this with SYCL_INTEL_bf16_conversion, and create a new experimental extension named "sycl_ext_intel_bfloat16" (or "sycl_ext_oneapi_bfloat16")? It seems like it would be easier for applications if all this bfloat16 functionality was tied to a single feature-test macro.
> 
> We also need to decide if these are "intel" or "oneapi" extensions. It makes no sense to have a "oneapi" extension layered on top of an "intel" extension.

Yes I think it would make sense to have a single bf16 extension as you suggest.  However since the current one is marked INTEL, I thought it made sense to add a separate proposal here for the time being.

I think that there only needs to be a single "oneapi" extension, and I already made an example PR that would switch the existing "intel" extension to support other backends, renaming it "oneapi", here: https://github.com/intel/llvm/pull/5393. In this proposal document here I referred to the INTEL named "SYCL_INTEL_bf16_conversion" rather than "SYCL_ONEAPI_bf16_conversion" simply because this is its current name, and I have not received an answer as to whether it would be acceptable to switch it to the "oneapi" namespace.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_bf16_math.asciidoc:
```diff
@@ -0,0 +1,105 @@
+
+## Introduction
+
+This document proposes extending the `fma`, `fmin`, `fmax` and `fabs` SYCL floating point math functions to support the `bfloat16` type introduced in the `SYCL_EXT_INTEL_BF16_CONVERSION` extension. This proposal assumes that devices which support the `SYCL_EXT_INTEL_BF16_CONVERSION` extension have the `bfloat16` scalar data type: `bfloat16`, and the `bfloat16` vector data types: `bfloat16_1`, `bf ...
```
I was told that I should initially propose the support of equivalent vector types to those that are currently supported in SYCL 2020 for the half case: this would mean that e.g. `bfloat16_2` is a vector of two `bfloat16`.  If we continue with the plan of implementing the corresponding vector types and these bfloat16 vector types are not defined in another extension then I imagine that we would define the vector types in this extension; but as you say it probably makes sense to have a single `bfloat16` extension. I think at this point we mainly need to know whether Intel also plans to support these functions at all with `bfloat16` in their backends or whether this extension would only apply to the CUDA backend at this moment in time? Further on from this if there is any idea on whether Intel would like to support bfloat16 vector types for these functions that would also be welcome.

I have asked whether the `aspect::ext_intel_bf16_conversion` could be implemented as a backend agnostic aspect here: https://github.com/intel/llvm/pull/5393#issuecomment-1042810593

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_bf16_math.asciidoc:
```diff
@@ -0,0 +1,105 @@
+# Bfloat16 math functions extension for DPC++: = SYCL_ONEAPI_bf16_math
```
The equals sign can be used similarly to # but not in the middle of the sentence. I removed `=`, thanks.

gmlueck commented:
I think you need to decide which `bfloat16_N` types are supported before this can be implemented, and you need to decide which extension provides those types.  If this extension only "proposed" at this time (meaning it's not implemented yet and could change), or is  it "experimental" (meaning it is already implemented)?  If it is only "proposed", the PR should be changed to put the file in that directory, and it is OK if these issues are not yet resolved.

We recently adopted a [template ](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc)to use for all new extensions.  Can you reformat this extension using that template?  See also the instructions in [README-process](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/README-process.md)

JackAKirk commented:
> I think you need to decide which `bfloat16_N` types are supported before this can be implemented, and you need to decide which extension provides those types. If this extension only "proposed" at this time (meaning it's not implemented yet and could change), or is it "experimental" (meaning it is already implemented)? If it is only "proposed", the PR should be changed to put the file in that directory, and it is OK if these issues are not yet resolved.

This extension has an initial implementation here: https://github.com/intel/llvm/pull/5748. In the same way as the matrix extension has temporary implementations using the uint16_t storage type of bfloat16 instead of bfloat16, this extension has temporary implementations of bfloat16 and bfloat16x2 using uint16_t and uint32_t respectively. The same applies for the bfloat16 part of the "fma_relu" extension: https://github.com/intel/llvm/pull/5749. As mentioned in the issues section of this design document the  `bfloat16_N` types that should be supported needs to be carefully considered. A first step is to switch the`bfloat16` class extension from Intel vendor specific to oneapi via https://github.com/intel/llvm/pull/5393 or otherwise.

> 
> We recently adopted a [template ](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc)to use for all new extensions. Can you reformat this extension using that template? See also the instructions in [README-process](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/README-process.md)

I'd prefer to first find consensus with Intel on whether this should be a separate extension or part of sycl_ext_*_bf16_conversion. If this is going to be a separate extension such that a separate design doc will be merged then I will correct the formatting.

JackAKirk commented:
> > I think you need to decide which `bfloat16_N` types are supported before this can be implemented, and you need to decide which extension provides those types. If this extension only "proposed" at this time (meaning it's not implemented yet and could change), or is it "experimental" (meaning it is already implemented)? If it is only "proposed", the PR should be changed to put the file in that directory, and it is OK if these issues are not yet resolved.
> 
> This extension has an initial implementation here: #5748. In the same way as the matrix extension has temporary implementations using the uint16_t storage type of bfloat16 instead of bfloat16, this extension has temporary implementations of bfloat16 and bfloat16x2 using uint16_t and uint32_t respectively. The same applies for the bfloat16 part of the "fma_relu" extension: #5749. We think that the `fma_relu` deserves a separate extension since it also covers `half` data types cases. As mentioned in the issues section of this design document the `bfloat16_N` types that should be supported needs to be carefully considered. A first step is to switch the`bfloat16` class extension from Intel vendor specific to oneapi via #5393 or otherwise.
> 
> > We recently adopted a [template ](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc)to use for all new extensions. Can you reformat this extension using that template? See also the instructions in [README-process](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/README-process.md)
> 
> I'd prefer to first find consensus with Intel on whether this should be a separate extension or part of sycl_ext_*_bf16_conversion. If this is going to be a separate extension such that a separate design doc will be merged then I will correct the formatting.

As discussed it has now been decided to have a single bfloat16 oneapi experimental extension.  As such the contents of the design document in this PR will be updated and merged with the bfloat16 oneapi experimental extension document once https://github.com/intel/llvm/pull/5393 is first merged.

JackAKirk commented:
Hi @gmlueck @dkhaldi 

This PR is now ready for review, along with the implementation in https://github.com/intel/llvm/pull/5964. I've added descriptions to the new functions in the extension document, and updated the "Issues" section. The function descriptions match the SYCL 2020 descriptions of the corresponding floating point math functions exactly except for fma: since "bfloat16" is not covered by the IEEE 754-2008 standard as far as I am aware. I've added test cases for correct NAN behaviour matching the descriptions of fmin and fmax to https://github.com/intel/llvm-test-suite/pull/975.

gmlueck reviewed (approved):
LGTM

@rdeodhar as FYI
